Modules/_sqlite/connection.c

```diff
@@ -323,6 +343,12 @@ pysqlite_connection_close_impl(pysqlite_Connection *self)
 /*[clinic end generated code: output=a546a0da212c9b97 input=3d58064bbffaa3d3]*/
 {
     int rc;
+    
+    if (!self->statements) {
+        PyErr_SetString(pysqlite_ProgrammingError,
+                        "Base Connection.__init__ not called.");
+        return NULL;
+    }
```

Comment:
Wouldn't it be better to handle this in `pysqlite_do_all_statements`, as it has multiple callers? Perhaps `pysqlite_do_all_statements` could be split in two: `do_all_statements` and `do_all_cursors`. Each of them would do a pointer sanity check before looping.

Reply:
Are you sure about splitting `pysqlite_do_all_statements` into two functions? It seems like it always performs the statements part, but optionally also performs the cursors part.

Reply:
No, I'm not sure :) Let's leave it as it is; both `pysqlite_do_all_statements` and this particular change.

Nit: `if (self->statements == NULL)` is more explicit, and I believe more consistent with the rest of the code base.